archive/imo/imo1998_q2.lean

```diff
@@ -0,0 +1,181 @@
+/-
+Copyright (c) 2020 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import data.fintype.basic
+import data.int.parity
+import algebra.big_operators.order
+import tactic.ring
+import tactic.noncomm_ring
+
+/-!
+# IMO 1998 Q2
+In a competition, there are `a` contestants and `b` judges, where `b ≥ 3` is an odd integer. Each
+judge rates each contestant as either "pass" or "fail". Suppose `k` is a number such that, for any
+two judges, their rating coincide for at most `k` contestants. Prove that `k / a ≥ (b - 1) / (2b)`.
+
+## Solution
+The problem asks us to think about triples consisting of a contestant and two judges whose ratings
+agree for that contestant. We thus consider the subset `A ⊆ C × JJ` of all such incidences of
+agreement, where `C` and `J` are the sets of contestants and judges, and `JJ = J × J − {(j, j)}`. We
+have natural maps: `left : A → C` and `right: A → JJ`. We count the elements of `A` in two ways: as
+the sum of the cardinalities of the fibres of `left` and as the sum of the cardinalities of the
+fibres of `right`. We obtain an upper bound on the cardinality of `A` from the count for `right`,
+and a lower bound from the count for `left`. These two bounds combine to the required result.
+
+First consider the map `right : A → JJ`. Since the size of any fibre over a point in JJ is bounded
+by `k` and since `|JJ| = b^2 - b`, we obtain the upper bound: `|A| ≤ k(b^2−b)`.
+
+Now consider the map `left : A → C`. The fibre over a given contestant `c ∈ C` is the set of
+ordered pairs of (distinct) judges who agree about `c`. We seek to bound the cardinality of this
+fibre from below. Minimum agreement for a contestant occurs when the judges' ratings are split as
+evenly as possible. Since `b` is odd, this occurs when they are divided into groups of size
+`(b−1)/2` and `(b+1)/2`. This corresponds to a fibre of cardinality `(b-1)^2/2` and so we obtain
+the lower bound: `a(b-1)^2/2 ≤ |A|`.
+
+Rearranging gives the result.
+-/
+
+open_locale classical
+noncomputable theory
+
+local notation x `/` y := (x : ℚ) / y
+
+variables {C J : Type*} (r : C → J → Prop)
+
+/-- The set of contestants on which two judges agree. -/
+def same_rating [fintype C] (j₁ j₂ : J) : finset C := finset.univ.filter (λ c, r c j₁ = r c j₂)
+
+section
+
+variables [fintype J] [fintype C]
+
+/-- All incidences of agreement. -/
+def A : finset (C × J × J) :=
+finset.univ.filter (λ (a : C × J × J), r a.1 a.2.1 = r a.1 a.2.2 ∧ a.2.1 ≠ a.2.2)
```

Comment:
I think that a `structure` will be more readable than `C × J × J`. If you don't want to use a `structure`, you can use `@[derive fintype] def triple := C × J × J` and add named projections `triple.contestant`, `triple.judge1`, `triple.judge2`.

Reply:
Agreed. I have just added a commit which adds named projections.

I did first experiment with using a `structure` but I eventually dismissed this approach as, at least with my implementation, it resulted in a fair amount of extra boiler plate.